### PR TITLE
LEAN_PATH as a mapping from package names to root dirs

### PR DESCRIPTION
**Pull request by Kha**
* change the semantics of `LEAN_PATH` as discussed
* implement all path handling in Lean
* no implicit `.Default` except for packages, i.e. zero lookup fallbacks
* relative imports are computed on module paths instead of file paths, ie. they cannot escape the current package

NOTE: After the next `update-stage0`, we will need to change `LEAN_PATH=.` to `LEAN_PATH=Init=.` at https://github.com/Kha/lean4/blob/4bf26193e5fbd0734c2d2f7ee8808a8563f34c4b/src/shell/CMakeLists.txt#L17

### CODE REVIEW

File: library/Init/Lean/Elaborator/Basic.lean [hunk 1]
```diff
@@ -386,12 +375,15 @@ match fileName with
 
 def testFrontend (input : String) (fileName : Option String := none) : IO (Environment × MessageLog) :=
 do env ← mkEmptyEnvironment;
-   baseDir ← toBaseDir fileName;
+   baseMod ← match fileName with
+   -- TODO: (try to) resolve current working directory to support relative imports with --stdin
```
Comment (leodemoura) [inline library/Init/Lean/Elaborator/Basic.lean]: I am happy to remove support for relative imports.

Comment (Kha) [reply library/Init/Lean/Elaborator/Basic.lean]: Yeah, I thought of that simplification too late and was surprised to find that relative imports are used exactly twice in mathlib (so probably an oversight). I will remove them.

**Comment by leodemoura** [comment]
> NOTE: After the next update-stage0, we will need to change LEAN_PATH=. to LEAN_PATH=Init=. at https://github.com/Kha/lean4/blob/4bf26193e5fbd0734c2d2f7ee8808a8563f34c4b/src/shell/CMakeLists.txt#L17

Tried `update-stage0` and applied this change. It didn't work for me. Build fails when trying to create `.depend` files.

```
/Users/leonardodemoura/projects/lean4/library/Init/Control/EState.lean:1:1: error: invalid import 'Init.Control.State', file './Control/State.olean' does not exist
/Users/leonardodemoura/projects/lean4/library/Init/Control/MonadFail.lean:1:1: error: invalid import 'Init.Control.Lift', file './Control/Lift.olean' does not exist
/Users/leonardodemoura/projects/lean4/library/Init/Control/Lift.lean:1:1: error: invalid import 'Init.Coe', file './Coe.olean' does not exist
/Users/leonardodemoura/projects/lean4/library/Init/WF.lean:1:1: error: invalid import 'Init.Data.Nat.Basic', file './Data/Nat/Basic.olean' does not exist
/Users/leonardodemoura/projects/lean4/library/Init/Control/Functor.lean:1:1: error: invalid import 'Init.Core', file './Core.olean' does not exist
/Users/leonardodemoura/projects/lean4/library/Init/Control/State.lean:1:1: error: /Users/leonardodemoura/projects/lean4/library/Init/Control/Alternative.lean/Users/leonardodemoura/projects/lean4/library/Init/Control/Conditional.lean::invalid import 'Init.Control.Alternative', file './Control/Alterna
```

**Comment by leodemoura** [comment]
@Kha I think the issue is the `--deps` option. Ignore `update-stage0`. If we just build this PR, execute `clean-olean`, and then try

```
~/projects/lean4/library (master)$ ../bin/lean --deps Init/Control/MonadFail.lean
/Users/leonardodemoura/projects/lean4/library/Init/Control/MonadFail.lean:1:1: error: invalid import 'Init.Control.Lift', file '/Users/leonardodemoura/projects/lean4/library/Init/Control/Lift.olean' does not exist
```

**Comment by Kha** [comment]
Thank you, this is ironically exactly what I said `--deps` should not do anymore: checking the filesystem. Previously it looked for a `.lean` _or_ `.olean` file, I think. I removed the check and tested `update-stage0`, it works with
```
--- a/src/shell/CMakeLists.txt
+++ b/src/shell/CMakeLists.txt
@@ -14,7 +14,7 @@ if(NOT STAGE0)
     # '-G Ninja' complains otherwise
     BYPRODUCTS "${CMAKE_BINARY_DIR}/stage1/libleanstdlib.a"
     WORKING_DIRECTORY "${LEAN_SOURCE_DIR}/../library"
-    COMMAND ${CMAKE_COMMAND} -E env "LEAN_PATH=." $(MAKE) "${CMAKE_BINARY_DIR}/stage1/libleanstdlib.a" "LEAN=$<TARGET_FILE:lean_stage0>" "LEANC_OPTS=${LEANC_OPTS}" "STAGE1_OUT=${CMAKE_BINARY_DIR}/stage1"
+    COMMAND ${CMAKE_COMMAND} -E env "LEAN_PATH=Init=Init" $(MAKE) "${CMAKE_BINARY_DIR}/stage1/libleanstdlib.a" "LEAN=$<TARGET_FILE:lean_stage0>" "LEANC_OPTS=${LEANC_OPTS}" "STAGE1_OUT=${CMAKE_BINARY_DIR}/stage1"
     DEPENDS stage0)
   set_target_properties(leanstdlib PROPERTIES
     IMPORTED_LOCATION "${CMAKE_BINARY_DIR}/stage1/libleanstdlib.a"
```
(so `Init=Init`, not `Init=.`)

**Comment by leodemoura** [comment]
Worked!

**Comment by leodemoura** [comment]
merged

**Pull request merged**